rllib/core/models/tests/test_base_models.py

```diff
@@ -0,0 +1,267 @@
+                        )
+                        return {"out_1": input_dict["in_1"], "out_2": out_2}
+
+            @dataclass
+            class CatModelConfig(ModelConfig):
+                def build(self, framework: str):
+                    # Since we define the correct model above anyway, we don't need
+                    # to distinguish between frameworks here.
+                    return TestModel(self)
+
+            # 1) Check if model behaves correctly with always_check_shapes=True first
+            # We expect model to raise an error if the input shapes are not correct.
+            # This is the behaviour we use for debugging with model specs.
+
+            config = CatModelConfig(always_check_shapes=True)
+
+            model = config.build(framework="spam")
+
+            # We want to raise an input spec validation error here since the input
+            # consists of lists and not torch Tensors
+            with self.assertRaisesRegex(ValueError, "input spec validation failed"):
+                model({"in_1": [1], "in_2": [1, 2]})
+
+            # We don't want to raise an input spec validation error here since the
+            # input consists of valid tensors
+            if fw == "torch":
+                model({"in_1": torch.Tensor([[1]]), "in_2": torch.Tensor([[1, 2]])})
+            else:
+                model({"in_1": tf.constant([[1]]), "in_2": tf.constant([[1, 2]])})
+
+            # 2) Check if model behaves correctly with always_check_shapes=False.
+            # We don't expect model to raise an error if the input shapes are not
+            # correct.
+            # This is the more performant default behaviour
+
+            config = CatModelConfig(always_check_shapes=False)
+
+            model = config.build(framework="spam")
+
+            # We want to raise an input spec validation error here since the input
+            # consists of lists and not torch Tensors
+            if fw == "torch":
+                model({"in_1": torch.Tensor([[1]]), "in_2": torch.Tensor([[1, 2]])})
+            else:
+                model({"in_1": tf.constant([[1]]), "in_2": tf.constant([[1, 2]])})
+            # We want to raise an input spec validation error here since the model
+            # itself does not raise an exception, so there is no need to double check
+            # the inputs.
+            if fw == "torch":
+                model(
+                    {"in_1": torch.Tensor([[1]]), "in_2": torch.Tensor([[1, 2, 3, 4]])}
+                )
+            else:
+                model({"in_1": tf.constant([[1]]), "in_2": tf.constant([[1, 2, 3, 4]])})
+
+            # We want to raise an input spec validation error here since the model
+            # raises an exception that stems from inputs that could have been caught
+            # with input spec checking.
+            with self.assertRaisesRegex(ValueError, "input spec validation failed"):
```

Comment:
Wait are you suggesting that I introduce a test that proves that spec checking takes time? 😄 
Do you think that's necessary? Because I already wrote a test that checks that there are no breaks for torch compile (torch dynamo test below)!

Reply:
that's right. I didn't see you had the torch compile test when writing this comment. I think that's sufficient then.